openmls/src/extensions/mod.rs

```diff
@@ -166,6 +170,180 @@ pub enum Extension {
+
+    /// Create an extension list that contains a single extension.
+    pub fn single(extension: Extension) -> Self {
+        Self {
+            map: HashMap::from([(extension.extension_type(), extension)]),
+        }
+    }
+
+    /// Create an extension list that contains multiple extensions.
+    ///
+    /// This function will fail when the list of extensions contains duplicate extension types.
+    pub fn multi(extensions: Vec<Extension>) -> Result<Self, InvalidExtensionError> {
+        extensions.try_into()
+    }
+
+    // ---------------------------------------------------------------------------------------------
+
+    /// Add an extension to the extension list.
+    ///
+    /// Returns an error when there already is an extension with the same extension type.
+    pub fn add(&mut self, extension: Extension) -> Result<(), InvalidExtensionError> {
+        if let Entry::Vacant(e) = self.map.entry(extension.extension_type()) {
+            e.insert(extension);
+            Ok(())
+        } else {
+            Err(InvalidExtensionError::Duplicate)
+        }
+    }
+
+    /// Add an extension to the extension list (or replace an existing one.)
+    ///
+    /// Returns the replaced extension (if any).
+    pub fn add_or_replace(&mut self, extension: Extension) -> Option<Extension> {
+        self.map.insert(extension.extension_type(), extension)
+    }
+
+    /// Remove an extension from the extension list.
+    ///
+    /// Returns the removed extension or `None` when there is no extension with the given extension type.
+    pub fn remove(&mut self, extension_type: ExtensionType) -> Option<Extension> {
+        self.map.remove(&extension_type)
+    }
+
+    /// Replace an extension in the extension list.
+    ///
+    /// Returns the replaced extension or an error when there is no extension with the given extension type.
+    #[cfg(any(feature = "test-utils", test))]
+    pub fn replace(&mut self, extension: Extension) -> Result<Extension, InvalidExtensionError> {
+        let got = self.map.remove(&extension.extension_type());
+
+        match got {
+            Some(extension) => Ok(extension),
+            None => Err(InvalidExtensionError::NotFound),
+        }
+    }
+}
+
+impl Default for Extensions {
+    fn default() -> Self {
+        Self::empty()
+    }
+}
+
+impl TryFrom<Vec<Extension>> for Extensions {
+    type Error = InvalidExtensionError;
+
+    fn try_from(candidate: Vec<Extension>) -> Result<Self, Self::Error> {
+        let mut map = HashMap::new();
+
+        for extension in candidate.into_iter() {
+            if map.insert(extension.extension_type(), extension).is_some() {
+                return Err(InvalidExtensionError::Duplicate);
+            }
+        }
+
+        Ok(Self { map })
+    }
+}
+
+impl Extensions {
+    /// Get a reference to the [`ApplicationIdExtension`] if there is any.
+    pub fn application_id(&self) -> Option<&ApplicationIdExtension> {
+        for extension in self.map.values() {
```

Comment:
There's no need to iterate here. We have the map. Something like this would work.
```rust
self.map.get(&ExtensionType::ApplicationId)
  .and_then(|e| match e {
    Extension::ApplicationId(e) => Some(e),
    _ => None,
})
```

Reply:
Oh yeah, thanks! Got a bit lazy...